pkg/discovery/worker/compliance/unschedulable_node_anti_affinity_group_dto_builder.go

```diff
@@ -0,0 +1,102 @@
+package compliance
+
+import (
+	"fmt"
+	"github.com/golang/glog"
+	"github.com/turbonomic/kubeturbo/pkg/discovery/repository"
+	"github.com/turbonomic/turbo-go-sdk/pkg/builder/group"
+	"github.com/turbonomic/turbo-go-sdk/pkg/proto"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+// In Turbo, Schedulable nodes are marked with an access commodity 'schedulable'. Pods are required to buy
+// this commodity so they can be moved to schedulable nodes
+// Unchedulable nodes do not sell this commodity, hence pods will not be moved to these nodes.
+// We want to address the issue that pods that are already running on the unschedulable nodes are not moved out
+// due to compliance issues. This is done by not requiring the pods already on the unschedulable nodes at the tine
+// of discovery to not buy the 'schedulable' commodity so they can continue staying on the current node unless there
+// are resource constraints or unless moved by the kubernetes controller.
+// In addition, we need pods which do not buy the schedulable commodity to not move to other unschedulable nodes.
+// This is achieved by creating an anti-affinity policy for these pods to the other unschedulable nodes other than the
+// one it is currently running on.
+type UnschedulableNodesAntiAffinityGroupDTOBuilder struct {
+	cluster  *repository.ClusterSummary
+	targetId string
+	// Manager for tracking schedulable nodes
+	nodesManager *SchedulableNodeManager
+}
+
+func NewUnschedulableNodesAntiAffinityGroupDTOBuilder(cluster *repository.ClusterSummary,
+	targetId string, nodesManager *SchedulableNodeManager) *UnschedulableNodesAntiAffinityGroupDTOBuilder {
+	return &UnschedulableNodesAntiAffinityGroupDTOBuilder{
+		cluster:      cluster,
+		targetId:     targetId,
+		nodesManager: nodesManager,
+	}
+}
+
+// Creates DTOs for anti-affinity policy between the pods running on a unschedulable nodes to other unschedulable nodes.
+// This will prevent the pods running on a unschedulable nodes to be moved to other other unschedulable nodes.
+func (builder *UnschedulableNodesAntiAffinityGroupDTOBuilder) Build() []*proto.GroupDTO {
+	var groupDTOs []*proto.GroupDTO
+
+	// Map of unschedulable node to other unschedulable nodes
+	unSchedulableNodesGroups := builder.getUnschedulableNodesGroups()
+
+	// Map of unschedulable node name to list of UIDs of pods running on these unschedulable nodes
+	nodeToPodsMap := builder.nodesManager.buildNodeToPodMap()
+
+	for unschedulableNodeName, nodeMembers := range unSchedulableNodesGroups {
```

Comment:
Suggestion: A better way of doing this can be:

Rename `unSchedulableNodes` to `unSchedulableNodesByUID` [here]( https://github.com/turbonomic/kubeturbo/blob/a863d96886732eca173a6e249fa13675cdbdc080/pkg/discovery/worker/compliance/schedulable_node_manager.go#L17:2)
Change [this line](https://github.com/turbonomic/kubeturbo/blob/a863d96886732eca173a6e249fa13675cdbdc080/pkg/discovery/worker/compliance/schedulable_node_manager.go#L59) to below:
`manager.unSchedulableNodesByUID = append(manager.unSchedulableNodesByUID, node.UID)`
Basically curate a list of node IDs in place of node names.

Then use below code.
```
       var nodeUIDtoNameMap make(map[string]string)
       var nodeUIDs []string
	for _, nodeName := range builder.nodesManager.unSchedulableNodes {
		nodeUID := builder.cluster.NodeNameUIDMap[nodeName]
               nodeUIDtoNameMap[nodeUID] = nodeName
		nodeUIDs = append(nodeUIDs, nodeUID)
	}

	for indx, nodeUID := range nodeUIDs {
    		var otherNodeUIDs []string
    		otherNodeUIDs := append(otherNodeUIDs, unSchedulableNodesByUID[:indx]...)
    		otherNodeUIDs := append(otherNodeUIDs, nodeUIDs[indx+1:]...)
    		nodeName := nodeUIDtoNameMap[nodeUID]

	// rest of your code below


	}
```